packages/arb-validator-core/gobridge/executionChallenge.go

```diff
@@ -0,0 +1,246 @@
+		totalSteps/uint64(bisectionCount)+totalSteps%uint64(bisectionCount),
+		preconditionHash,
+		assertionHash,
+	))
+
+	for i := 1; i < bisectionCount; i++ {
+		if didInboxInsns[i-1] {
+			precondition.BeforeInbox = value.NewEmptyTuple()
+		}
+		assertionHash = generateAssertionHash(
+			machineHashes[i+1],
+			didInboxInsns[i],
+			gasses[i],
+			messageAccs[i],
+			messageAccs[i+1],
+			logAccs[i],
+			logAccs[i+1],
+		)
+		hashes = append(hashes, valprotocol.ExecutionDataHash(
+			totalSteps/uint64(bisectionCount),
+			valprotocol.ExecutionPreconditionHash(machineHashes[i], precondition.TimeBounds, precondition.BeforeInbox.Hash()),
+			assertionHash))
+	}
+
+	c.commitToSegment(hashes)
+	c.asserterResponded()
+
+	c.client.GoEthClient.pubMsg(c.challengeData, arbbridge.MaybeEvent{
+		Event: arbbridge.ExecutionBisectionEvent{
+			ChainInfo: arbbridge.ChainInfo{
+				BlockId: c.client.GoEthClient.getCurrentBlock(),
+			},
+			Assertions: assertions,
+			TotalSteps: totalSteps,
+			Deadline:   c.client.GoEthClient.challenges[c.contractAddress].deadline,
+		},
+	})
+
+	return nil
+}
+func hashSliceToHashes(slice [][32]byte) []common.Hash {
+	ret := make([]common.Hash, 0, len(slice))
+	for _, a := range slice {
+		ret = append(ret, a)
+	}
+	return ret
+}
+
+func (c *ExecutionChallenge) OneStepProof(
+	ctx context.Context,
+	precondition *valprotocol.Precondition,
+	assertion *valprotocol.ExecutionAssertionStub,
+	proof []byte,
+) error {
+	valprotocol.ExecutionPreconditionHash(precondition.BeforeHash, precondition.TimeBounds, precondition.BeforeInbox.Hash())
+	precondition.Hash()
+
+	matchHash := valprotocol.ExecutionDataHash(1, precondition.Hash(), assertion.Hash())
+	if !c.client.GoEthClient.challenges[c.contractAddress].challengerDataHash.Equals(matchHash) {
+		return errors.New("OneStepProof Incorrect previous state")
+	}
+
+	// TODO: executionChallenge one step proof validation
+	//	uint256 correctProof = OneStepProof.validateProof(
+	//		_beforeHash,
+	//		_timeBoundsBlocks,
+	//		_beforeInbox,
+	//		_afterHash,
+	//		_didInboxInsns,
+	//		_firstMessage,
+	//		_lastMessage,
+	//		_firstLog,
+	//		_lastLog,
+	//		_gas,
+	//		_proof
+	//	);
+	//
+	// for now make OSP always valid
+
+	//	require(correctProof == 0, OSP_PROOF);
+	//	emit OneStepProofCompleted();
+	c.client.GoEthClient.pubMsg(c.challengeData, arbbridge.MaybeEvent{
+		Event: arbbridge.OneStepProofEvent{
+			ChainInfo: arbbridge.ChainInfo{
+				BlockId: c.client.GoEthClient.getCurrentBlock(),
+			},
+		},
+	})
+
+	//c.client.GoEthClient.pubMsg(arbbridge.MaybeEvent{
+	//	Event: arbbridge.ChallengeCompletedEvent{
+	//		ChainInfo: arbbridge.ChainInfo{
+	//			BlockId: c.client.GoEthClient.getCurrentBlock(),
+	//		},
+	//		Winner:            c.client.GoEthClient.challenges[c.contractAddress].asserter,
+	//		Loser:             c.client.GoEthClient.challenges[c.contractAddress].challenger,
+	//		ChallengeContract: c.contractAddress,
+	//	},
+	//})
+
+	//	_asserterWin();
```

Comment:
The challenge currently doesn't seem to call the rollup chain to tell it the challenge is resolved.

Reply:
implemented resolveChallenge